File: ports/java/src/core/kernel/BcsObjectCleanupHandler.java
package core.kernel;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.locks.ReentrantLock;

public class BcsObjectCleanupHandler extends BcsObject {
    private final List<BcsObject> cleanupObjects = new ArrayList<>();
    private final ReentrantLock lock = new ReentrantLock();

    public BcsObjectCleanupHandler() {
        super(null);
    }

    public BcsObject add(BcsObject obj) {
        lock.lock();
        try {
            if (!cleanupObjects.contains(obj)) {
                cleanupObjects.add(obj);
            }
            return obj;
        } finally {
            lock.unlock();
        }
    }

    public void remove(BcsObject obj) {
        lock.lock();
        try {
            cleanupObjects.remove(obj);
        } finally {
            lock.unlock();
        }
    }

    public boolean isEmpty() {
        lock.lock();
        try {
            return cleanupObjects.isEmpty();
        } finally {
            lock.unlock();
        }
    }

    public void clear() {
        List<BcsObject> objectsToClear;
        lock.lock();
        try {
            objectsToClear = new ArrayList<>(cleanupObjects);
            cleanupObjects.clear();
        } finally {
            lock.unlock();
        }

        for (BcsObject obj : objectsToClear) {
            obj.close();
        }
    }
}

File: ports/csharp/src/core/kernel/BcsObjectCleanupHandler.cs
using System.Collections.Generic;

namespace Bcs.Core.Kernel
{
    public class BcsObjectCleanupHandler : BcsObject
    {
        private readonly List<BcsObject> _cleanupObjects = new List<BcsObject>();
        private readonly object _lock = new object();

        public BcsObjectCleanupHandler() : base(null)
        {
        }

        public BcsObject Add(BcsObject obj)
        {
            lock (_lock)
            {
                if (!_cleanupObjects.Contains(obj))
                {
                    _cleanupObjects.Add(obj);
                }
            }
            return obj;
        }

        public void Remove(BcsObject obj)
        {
            lock (_lock)
            {
                _cleanupObjects.Remove(obj);
            }
        }

        public bool IsEmpty()
        {
            lock (_lock)
            {
                return _cleanupObjects.Count == 0;
            }
        }

        public void Clear()
        {
            List<BcsObject> objectsToClear;
            lock (_lock)
            {
                objectsToClear = new List<BcsObject>(_cleanupObjects);
                _cleanupObjects.Clear();
            }

            foreach (var obj in objectsToClear)
            {
                obj.Dispose();
            }
        }
    }
}
